worker/upload.go

```diff
@@ -370,6 +391,58 @@ func (mgr *uploadManager) numUploaders() int {
 	return len(mgr.uploaders)
 }
 
+func (mgr *uploadManager) uploader(shard *shardUpload) *uploader {
+	mgr.mu.Lock()
+	if len(mgr.uploaders) == 0 {
+		mgr.mu.Unlock()
+		return nil
+	}
+
+	// sort the uploaders by their estimate
+	sort.Slice(mgr.uploaders, func(i, j int) bool {
+		return mgr.uploaders[i].estimate() < mgr.uploaders[j].estimate()
+	})
+
+	// select top ten candidates
+	var candidates []*uploader
+	for _, uploader := range mgr.uploaders {
+		if shard.upload.canUseUploader(uploader, shard.sID) {
+			candidates = append(candidates, uploader)
+			if len(candidates) == 10 {
+				break
+			}
+		}
+	}
+	mgr.mu.Unlock()
+
+	// return early if we have no queues left
+	if len(candidates) == 0 {
+		return nil
+	}
+
+loop:
+	for {
+		// if this slab does not have more than 1 parent, we resort the
+		// candidates and return the best one at the time
+		if len(shard.upload.parents(shard.sID)) <= 1 {
```

Comment:
(nit) the candidates are sorted the first time you arrive here right? So how about only doing the 
```go
if len(shard.upload.parents(shard.sID)) <= 1 {
  return candidates[0]
}
```
here and move the sorting down to under the select?